### PR TITLE
Well-founded recursion without infinity

### PR DESCRIPTION
**Pull request by sctfn**
Continuation of #4290. This MR brings in well-founded recursion over a partial order. This can be used to revise wrecs and shorten a lot of theorems there.

### CODE REVIEW

File: set.mm
Comment (benjub) [inline set.mm:79697]: This (and previous and next statements) can use fewer lines while retaining sensible line-breaks.

File: set.mm
Comment (benjub) [inline set.mm:79799]:
```suggestion
       is a relation.  (Contributed by Scott Fenton, 27-Aug-2022.) $)
```

File: set.mm
Comment (benjub) [inline set.mm:80074]: Here and above/below, I'd write "functionality".

**Review by benjub — COMMENTED** [review]
A few remarks. That's a very nice addition to our fundamental constructions.

File: set.mm
Comment (benjub) [reply set.mm:79651]: (And that particular theorem is not a very good example, since the need of deduction form would be very rare: we have the choice to use a fresh variable x; for other theorems, there may be a need.)

**Comment by benjub** [comment]
Also: have you had a look at BNJ's mathbox ? If the work is related, we should find a place to add some credit to him.

File: set.mm
Comment (sctfn) [reply set.mm]: Done

File: set.mm
Comment (sctfn) [reply set.mm:80074]: Done

**Comment by sctfn** [comment]
Sorry, who's BNJ? You're the only BJ I can find in set.mm

**Comment by benjub** [comment]
> Sorry, who's BNJ? You're the only BJ I can find in set.mm

Sorry, I meant "Jonathan Ben-Naim". Most of his theorems are labeled `bnj*`, hence my confusion.  No relation whatsoever with me.  See https://us.metamath.org/mpeuni/mmtheorems.html#dtl:20.4

He's apparently a former contributor and his theorems date from before I even know of metamath. His mathbox has sections on recursion too.

**Comment by sctfn** [comment]
Hm, he and I seem to have developed the same theorems around the same time. We even used the same reference! His theorems only cover the case that requires infinity though. I'm not completely sure what to do about credit here.

**Comment by benjub** [comment]
> Hm, he and I seem to have developed the same theorems around the same time. We even used the same reference! His theorems only cover the case that requires infinity though. I'm not completely sure what to do about credit here.

Thanks for looking at that. We'll find a way to credit him, maybe in the head comment of the section.

**Review by benjub — APPROVED** [review]

**Comment by sctfn** [comment]
Can I get a second review here? I'd feel better merging with another pair of eyes.

File: set.mm
Comment (icecream17) [inline set.mm:80327]:
```suggestion
    $( The well-ordered recursion generator generates a relation.
```

for consistency I guess

File: set.mm
Comment (icecream17) [reply set.mm:79614]: it doesn't look like these suggestions were changed, so I assume they're done in a future pr

File: set.mm
Comment (icecream17) [reply set.mm:79617]: Perhaps it would be simpler to state "The proof requires either a partial order or infinity" or something, like the comment of ~ frrlem9 (this pr). This is optional

**Review by icecream17 — APPROVED** [review]

File: set.mm
Comment (sctfn) [reply set.mm:79614]: Huh, looks like the github app didn't commit these changes. I'll do it the normal way.

**Pull request merged**
